packages/middleware-bucket-endpoint/src/bucketHostnameUtils.ts

```diff
@@ -0,0 +1,128 @@
+import { ARN } from "@aws-sdk/util-arn-parser";
+
+const DOMAIN_PATTERN = /^[a-z0-9][a-z0-9\.\-]{1,61}[a-z0-9]$/;
+const IP_ADDRESS_PATTERN = /(\d+\.){3}\d+/;
+const DOTS_PATTERN = /\.\./;
+export const DOT_PATTERN = /\./;
+export const S3_HOSTNAME_PATTERN = /^(.+\.)?s3[.-]([a-z0-9-]+)\./;
+const S3_US_EAST_1_ALTNAME_PATTERN = /^s3(-external-1)?\.amazonaws\.com$/;
+const AWS_PARTITION_SUFFIX = "amazonaws.com";
+
+export interface AccessPointArn extends ARN {
+  accessPointName: string;
+}
+
+export interface BucketHostnameParameters {
+  baseHostname: string;
+  bucketName: string;
+  accelerateEndpoint?: boolean;
+  dualstackEndpoint?: boolean;
+  pathStyleEndpoint?: boolean;
+  tlsCompatible?: boolean;
+}
+
+export interface ArnHostnameParameters extends Omit<BucketHostnameParameters, "bucketName"> {
+  bucketName: ARN;
+  clientSigningRegion?: string;
+  clientPartition?: string;
+  useArnRegion?: boolean;
+}
+
+export const isBucketNameOptions = (
+  options: BucketHostnameParameters | ArnHostnameParameters
+): options is BucketHostnameParameters => typeof options.bucketName === "string";
+
+export const getPseudoRegion = (region: string) => (isFipsRegion(region) ? region.replace(/fips-|-fips/, "") : region);
+
+/**
+ * Determines whether a given string is DNS compliant per the rules outlined by
+ * S3. Length, capitaization, and leading dot restrictions are enforced by the
+ * DOMAIN_PATTERN regular expression.
+ *
+ * @see https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html
+ */
+export const isDnsCompatibleBucketName = (bucketName: string): boolean =>
+  DOMAIN_PATTERN.test(bucketName) && !IP_ADDRESS_PATTERN.test(bucketName) && !DOTS_PATTERN.test(bucketName);
+
+const getRegionalSuffix = (hostname: string): [string, string] => {
+  const parts = hostname.match(S3_HOSTNAME_PATTERN)!;
+  return [parts[2], hostname.replace(new RegExp(`^${parts[0]}`), "")];
+};
+
+export const getSuffix = (hostname: string): [string, string] =>
+  S3_US_EAST_1_ALTNAME_PATTERN.test(hostname) ? ["us-east-1", AWS_PARTITION_SUFFIX] : getRegionalSuffix(hostname);
+
+export const getSuffixForArnEndpoint = (hostname: string): [string, string] =>
+  S3_US_EAST_1_ALTNAME_PATTERN.test(hostname)
+    ? [hostname.replace(`.${AWS_PARTITION_SUFFIX}`, ""), AWS_PARTITION_SUFFIX]
+    : getRegionalSuffix(hostname);
+
+export const validateService = (service: string) => {
+  if (service !== "s3") {
+    throw new Error("Expect 's3' in access point ARN service component");
+  }
+};
+
+export const validatePartition = (partition: string, options: { clientPartition: string }) => {
+  if (partition !== options.clientPartition) {
+    throw new Error(`Partition in ARN is incompatible, got ${partition} but expected ${options.clientPartition}`);
```

Comment:
nit: put quotes around data?
```suggestion
    throw new Error(`Partition in ARN is incompatible, got "${partition}" but expected "${options.clientPartition}"`);
```

Ditto for line 88, 115